include/gridtools/reduction/gpu.hpp

```diff
@@ -0,0 +1,318 @@
+                constexpr size_t max_threads = 1024;
+                assert(threads >= 1);
+                assert(threads <= max_threads);
+                ct_dispatch<int_log2(max_threads) + 1>(
+                    [=](auto n) {
+                        constexpr size_t block_size = 1 << decltype(n)::value;
+                        using kernel_t = choose_kernel<block_size, T>;
+                        launch<<<blocks, threads>>>(kernel_t(), f, in, out);
+                        GT_CUDA_CHECK(cudaGetLastError());
+                    },
+                    int_log2(threads));
+            }
+
+            template <class T>
+            inline unsigned char const *memset_cast(T const &val) {
+                auto res = reinterpret_cast<unsigned char const *>(&val);
+                T buff;
+                return std::memcmp(&val, std::memset(&buff, *res, sizeof(T)), sizeof(T)) == 0 ? res : nullptr;
+            }
+
+            struct gpu {};
+
+            template <class F, class T>
+            auto reduction_reduce(gpu, F f, T *ptr, size_t size) {
+                T *in = ptr;
+                size_t n = size;
+                while (n > cpu_final_threshold_t::value) {
+                    size_t threads = get_threads(n);
+                    size_t blocks = n / (2 * threads);
+                    T *out = in + n;
+                    reduce_step(f, threads, blocks, in, out);
+                    n = blocks;
+                    in = out;
+                }
+                T buff[cpu_final_threshold_t::value];
+                GT_CUDA_CHECK(cudaMemcpy(buff, in, n * sizeof(T), cudaMemcpyDeviceToHost));
+                return reduce_cpu(f, buff, n);
+            }
+
+            inline size_t reduction_round_size(gpu, size_t size) {
+                auto chunk = next_pow2(size) / cpu_final_threshold_t::value;
+                return chunk ? (size + chunk - 1) / chunk * chunk : size;
+            }
+
+            inline size_t reduction_allocation_size(gpu, size_t n) {
+                size_t res = 0;
+                do {
+                    res += n;
+                    n /= 2 * get_threads(n);
+                } while (n > cpu_final_threshold_t::value);
+                res += n;
+                return res;
+            }
+
+            template <class T>
+            void reduction_fill(gpu,
+                T const &initial_value,
+                T *dst,
+                size_t data_size,
+                size_t rounded_size,
+                size_t /*allocation_size*/,
+                bool has_holes) {
+                if (!has_holes) {
+                    dst += data_size;
+                    rounded_size -= data_size;
+                }
+                if (!rounded_size)
+                    return;
+                size_t bytes = rounded_size * sizeof(T);
+                if (auto init = memset_cast(initial_value)) {
```

Comment:
I see that there are just two cases where this succeeds: either all bits are zero or the bit-pattern repeats every byte (which in practice probably happens exactly when all bits are zero). I would assume that the difference between a call to `cudaMemset` and a trivial kernel that just sets `ptr[i] = initial_value;` (see comment below) is actually negligible and could simplify the code path here.

Reply:
There is one more practical case -- max value for unsigned and min value for signed integers. And it is a good idea for the dedicated kernel!

Reply:
done